common/src/Renderer/GLVertexAttributeType.h

```diff
@@ -20,296 +20,185 @@
+            // EntityLinkArrow
+            arrowPosition,
+            lineDir
+        };
 
-            /**
-             * Sets up a vertex buffer pointer for this attribute with the given index, stride, and offset.
-             *
-             * @param index the attribute's index (position in the vertices)
-             * @param stride the stride for the vertex buffer pointer
-             * @param offset the offset for the vertex buffer pointer
-             */
-            static void setup(const size_t index, const size_t stride, const size_t offset) {
-                unused(index);
-                unused(stride);
-                unused(offset);
+        inline const char* toString(const AttributeNames name) {
+            switch (name) {
+                case AttributeNames::arrowPosition: return "arrowPosition";
+                case AttributeNames::lineDir: return "lineDir";
+                switchDefault()
             }
-            static void cleanup(const size_t /* index */) {}
-
-            // Non-instantiable
-            GLVertexAttributeType() = delete;
-            deleteCopyAndMove(GLVertexAttributeType)
-        };
+        }
 
         /**
-         * Partial specialization for user defined vertex attribute types.
+         * User defined vertex attribute types.
          *
+         * @tparam A the attribute name
          * @tparam D the vertex component type
          * @tparam S the number of components
+         * @tparam N whether to convert signed integer types to [-1..1] and unsigned to [0..1]
          */
-        template <GLenum D, size_t S>
-        class GLVertexAttributeType<GLVertexAttributeTypeTag::User, D, S> {
+        template <AttributeNames A, GLenum D, size_t S, Normalize N>
+        class GLVertexAttributeUser {
         public:
             using ComponentType = typename GLType<D>::Type;
             using ElementType = vm::vec<ComponentType,S>;
             static const size_t Size = sizeof(ElementType);
 
-            static void setup(const size_t index, const size_t stride, const size_t offset) {
-                glAssert(glEnableVertexAttribArray(static_cast<GLuint>(index)))
-                glAssert(glVertexAttribPointer(static_cast<GLuint>(index), static_cast<GLint>(S), D, 0, static_cast<GLsizei>(stride), reinterpret_cast<GLvoid*>(offset)))
+            static void setup(ShaderProgram* program, const size_t /* index */, const size_t stride, const size_t offset) {
+                ensure(program != nullptr, "must have a program bound to use generic attributes");
+
+                const GLint attributeIndex = program->findAttributeLocation(toString(A));
+                glAssert(glEnableVertexAttribArray(static_cast<GLuint>(attributeIndex)))
+                glAssert(glVertexAttribPointer(static_cast<GLuint>(attributeIndex), static_cast<GLint>(S), D, (N == Normalize::True) ? GL_TRUE : GL_FALSE, static_cast<GLsizei>(stride), reinterpret_cast<GLvoid*>(offset)))
```

Comment:
Instead of passing this as a template parameter, could you not just overload `setup` and add the additional parameter? That would also allow us to keep the original partical specialization design. I think that would be more coherent because `normalize` is just another parameter of `glVertexAttribPointer`, of which there are multiple overloads as well.

Reply:
It needs to be part of the type, but if we kept the old design what we could do is have:
GLVertexAttributeTypeTag::UserUnnormalized andGLVertexAttributeTypeTag::UserNormalized

(what this does is, if your data type is GL_UNSIGNED_BYTE, it controls whether the shader sees that mapped to a float in the range 0...127 or 0..1)

But I don't understand the appeal of the old code with the enum, when I found out it could be refactored like this I thought I had made a major cleanup.

Reply:
Or at least, a bit of a cleanup :)

But I realize the old approach still wouldn't work, because we need to include the string name of the attribute as a template param, so GLVertexAttributeUser would need one extra param compared with the others anyway.

Reply:
If we go with my suggestion of a type that carries the attribute name, we can also pass the normalization parameter as a boolean template parameter. But I'm not sure if we don't want the normaliation parameter for other vertex attribute types as well, in which case I would still vote for an overloaded `setup` method.

Reply:
> If we go with my suggestion of a type that carries the attribute name, we can also pass the normalization parameter as a boolean template parameter. 

Yeah, that sounds fine. Shouldn't it also have the datatype and number of components too, then?

The normalized param was a new thing in OpenGL for generic attributes, I don't think any of the old fixed-function attributes let you choose normalized or not. I assume glNormalPointer would be normalized, and the other ones probably not?

Reply:
> I would still vote for an overloaded `setup` method.

I'm not seeing how it would work, because the code that calls `setup` doesn't know which class it's going to be calling `setup` on, so all the setup functions need to have the same signature.

Reply:
Ah, I see. Ok, then it makes sense to pass it as a template parameter.

From the OpenGL manual:

> For glVertexAttribPointer, if normalized is set to GL_TRUE, it indicates that values stored in an integer format are to be mapped to the range [-1,1] (for signed values) or [0,1] (for unsigned values) when they are accessed and converted to floating point. Otherwise, values will be converted to floats directly without normalization. 

So I'm not sure if this has anything to do with normals. I also don't know what it means by "mapped". How are these values mapped? Are they clamped? Scaled? No idea. What do you  need it for?

Reply:
> I also don't know what it means by "mapped". How are these values mapped? Are they clamped? Scaled? No idea. 

There's no clamping, it's about how integers in your vertex bufffers are converted to floats before the shader sees them. So if the datatype is GL_UNSIGNED_BYTE and normalized is set to GL_TRUE, the mapping is 0->0.0 and 255->1.0. If it's set to false, the mapping is 0->0.0, and 255->255.0.

For example, we should be using 1 byte (GL_UNSIGNED_BYTE  / uint8_t) per components anywhere we pass colors per vertex; currently we're using 4 byte (float) per component. In this case we'd probably want to set normalize to GL_TRUE, so the shader sees color values that are already in the range 0..1.

Reply:
Ok, thanks for explaining. In that case, I request only the following changes:

- pass the attribute name as a struct with a static string member
- pass the normalize parameter as a boolean template parameter instead of an enum

Does that make sense?

Reply:
Sounds good!